external-crates/move/crates/move-package/src/compilation/compiled_package.rs

```diff
@@ -756,3 +765,123 @@ pub(crate) fn make_source_and_deps_for_compiler(
+                deps_for_current_compiler.push(dep)
+            }
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == env!("CARGO_PKG_VERSION") => {
+                // This dependency requires the same compiler version we're on.
+                deps_for_current_compiler.push(dep)
+            }
+            Some(toolchain_version) => {
+                // Compile and mark that we compiled this dep with a prior compiler.
+                download_and_compile(root, toolchain_version)?;
+                deps_for_prior_compiler.push(dep)
+            }
+        }
+    }
+    Ok((deps_for_current_compiler, deps_for_prior_compiler))
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: ToolchainVersion,
+) -> Result<()> {
+    let binaries_path = std::env::var("TOOLCHAIN_BINARIES")?; // E.g., ~/.move/binaries
+    let dest_dir = Path::new(binaries_path.as_str());
+    let dest_version = dest_dir.join(compiler_version.clone());
+    let platform = "macos-arm64"; // Hardcoded platform: reviewer please ignore this for now.
+    let dest_binary = dest_version.join(format!("target/release/sui-{}", platform));
+    let dest_binary_os = OsStr::new(dest_binary.as_path());
+
+    if !dest_binary.exists() {
+        // Download if binary does not exist.
+        let url = format!("https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz", compiler_version, compiler_version, platform);
+        let release_url = OsStr::new(url.as_str());
+        let dest_tarball = dest_version.join(format!("{}.tgz", compiler_version));
+
+        println!(
+            "[+] curl -L --create-dirs -o {} {}",
+            dest_tarball.display(),
+            url,
+        );
+        let result = Command::new("curl")
+            .args([
+                OsStr::new("-L"),
+                OsStr::new("--create-dirs"),
+                OsStr::new("-o"),
+                OsStr::new(dest_tarball.as_path()),
+                OsStr::new(release_url),
+            ])
+            .output()
+            .expect("failed to download");
+
+        println!(
+            "[+] tar -xzf {} -C {}",
+            dest_tarball.display(),
+            dest_version.display()
+        );
+        let result = Command::new("tar")
+            .args([
+                OsStr::new("-xzf"),
+                OsStr::new(dest_tarball.as_path()),
+                OsStr::new("-C"),
+                OsStr::new(dest_version.as_path()),
+            ])
+            .output()
+            .expect("failed to untar");
+
+        let mut perms = fs::metadata(dest_binary_os)?.permissions();
+        perms.set_mode(0o755);
+        fs::set_permissions(dest_binary_os, perms)?;
+    }
+
+    println!(
+        "[+] sui move build --default-move-edition {} --default-move-flavor {} -p {}",
```

Comment:
I worry about these flags here just in case we accidentally break them. I mean... we shouldn't but we might. Thoughts?

Reply:
IMO flag shifting is a real possibility. Perhaps a file that defines invocations per-version?

Reply:
We'll need (a different?) way to build if we can't rely on flags--I'm not aware of another way right now if we're invoking the binaries? I'd love if we can have a high degree of confidence about flags, at least for medium to long term (6+ months?). That way, if they do change, one way to approach this is to have version detection logic added in new releases, that case out and use appropriate flags. Basically, build it into this logic.

> Perhaps a file that defines invocations per-version?

Proposal for what this would look like? (Will this mean another file users need to check in, and do we really want that?)